proto/cometbft/abci/v4/types.proto

```diff
@@ -0,0 +1,233 @@
+syntax = "proto3";
+package cometbft.abci.v4;
+
+import "cometbft/abci/v1/types.proto";
+import "cometbft/abci/v2/types.proto";
+import "cometbft/abci/v3/types.proto";
+import "gogoproto/gogo.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/cometbft/cometbft/api/cometbft/abci/v4";
+
+service ABCI {
+  rpc Echo(v1.RequestEcho) returns (v1.ResponseEcho);
+  rpc Flush(v1.RequestFlush) returns (v1.ResponseFlush);
+  rpc Info(v2.RequestInfo) returns (v1.ResponseInfo);
+  rpc CheckTx(RequestCheckTx) returns (v3.ResponseCheckTx);
+  rpc Query(v1.RequestQuery) returns (v1.ResponseQuery);
+  rpc Commit(v1.RequestCommit) returns (v3.ResponseCommit);
+  rpc InitChain(v3.RequestInitChain) returns (v3.ResponseInitChain);
+  rpc ListSnapshots(v1.RequestListSnapshots) returns (v1.ResponseListSnapshots);
+  rpc OfferSnapshot(v1.RequestOfferSnapshot) returns (ResponseOfferSnapshot);
+  rpc LoadSnapshotChunk(v1.RequestLoadSnapshotChunk) returns (v1.ResponseLoadSnapshotChunk);
+  rpc ApplySnapshotChunk(v1.RequestApplySnapshotChunk) returns (ResponseApplySnapshotChunk);
+  rpc PrepareProposal(RequestPrepareProposal) returns (v2.ResponsePrepareProposal);
+  rpc ProcessProposal(RequestProcessProposal) returns (ResponseProcessProposal);
+  rpc ExtendVote(v3.RequestExtendVote) returns (v3.ResponseExtendVote);
+  rpc VerifyVoteExtension(v3.RequestVerifyVoteExtension) returns (ResponseVerifyVoteExtension);
+  rpc FinalizeBlock(RequestFinalizeBlock) returns (v3.ResponseFinalizeBlock);
+}
+
+//----------------------------------------
+// Request types
+
+message Request {
+  oneof value {
+    v1.RequestEcho echo = 1;
+    v1.RequestFlush flush = 2;
+    v2.RequestInfo info = 3;
+    v3.RequestInitChain init_chain = 5;
+    v1.RequestQuery query = 6;
+    RequestCheckTx check_tx = 8;
+    v1.RequestCommit commit = 11;
+    v1.RequestListSnapshots list_snapshots = 12;
+    v1.RequestOfferSnapshot offer_snapshot = 13;
+    v1.RequestLoadSnapshotChunk load_snapshot_chunk = 14;
+    v1.RequestApplySnapshotChunk apply_snapshot_chunk = 15;
+    RequestPrepareProposal prepare_proposal = 16;
+    RequestProcessProposal process_proposal = 17;
+    v3.RequestExtendVote extend_vote = 18;
+    v3.RequestVerifyVoteExtension verify_vote_extension = 19;
+    RequestFinalizeBlock finalize_block = 20;
+  }
+  reserved 4, 7, 9, 10; // SetOption, BeginBlock, DeliverTx, EndBlock
+}
+
+enum CheckTxType {
+  option (gogoproto.goproto_enum_prefix) = false;
+
+  // The non-conventional name for the zero value is preserved
+  // due to history of use.
+  // buf:lint:ignore ENUM_ZERO_VALUE_SUFFIX
+  CHECK_TX_TYPE_NEW = 0;
+  CHECK_TX_TYPE_RECHECK = 1;
+}
```

Comment:
This would be a clean, if disruptive, way to fix the forward compatibility issue.
With domain types e.g. in tendermint-rs, we can also convert seamlessly between the old and the new encodings.

Reply:
I have regularized `CheckTxType` in the proposed way, and changed the fields of `RequestCheckTx` to ensure the clean break.